src/filesystem_helper.cpp

```diff
@@ -306,6 +306,40 @@ path temp_directory_path()
   return path(temp_path);
 }
 
+path create_temp_directory(const std::string & base_name, const path & parent_path)
+{
+  const auto template_path = base_name + "XXXXXX";
+  std::string full_template_str = (parent_path / template_path).string();
+  char * mutable_full_template_data = &full_template_str[0];
```

Comment:
The problem with `char * mutable_full_template_data = &full_template_str[0];` is the assumption it makes about how `std::string` stores its data. Something like:
```suggestion
  std::vector<char> mutable_full_template_data(full_template_str.begin(), full_template_str.end());
  // now use mutable_full_template_data.data()
```
doesn't have that issue.

Reply:
I don't particularly care which way we go, but I will point out that the [documentation](https://en.cppreference.com/w/cpp/string/basic_string) for string says:

```
The elements of a basic_string are stored contiguously, that is, for a basic_string s, &*(s.begin() + n) == &*s.begin() + n for any n in [0, s.size()), or, equivalently, a pointer to s[0] can be passed to functions that expect a pointer to the first element of a null-terminated (since C++11)CharT[] array.
```

Reply:
Uh, I stand corrected then. Thanks @clalancette. Not sure why I was convinced the standard did not enforce contiguity.

Reply:
Ok - I'll change it back, for performance reasons. This implementation has to perform two full copies that aren't necessary if we just go the `&string[0]` route